azure/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,141 @@
+)
+
+// NatGatewayScope defines the scope interface for nat gateway service
+type NatGatewayScope interface {
+	logr.Logger
+	azure.ClusterScoper
+	NatGatewaySpecs() []azure.NatGatewaySpec
+	SetNodeNatGateway(natGateway infrav1.NatGateway)
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope NatGatewayScope
+	client
+}
+
+// New creates a new service.
+func New(scope NatGatewayScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile gets/creates/updates a nat gateway.
+// Only when the Nat Gateway 'Name' property is defined we create the Nat Gateway: it's opt-in. Should we use the nat gateway ip name instead?
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
+
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
+
+		switch {
+		case err != nil && !azure.ResourceNotFound(err):
+			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+		case err == nil:
+			// nat gateway already exists
+			s.Scope.V(2).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
+			s.Scope.SetNodeNatGateway(*existingNatGateway)
+		default:
+			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
+			s.Scope.V(2).Info("creating nat gateway", "nat gateway", natGatewaySpec.Name)
+			err = s.client.CreateOrUpdate(
+				ctx,
+				s.Scope.ResourceGroup(),
+				natGatewaySpec.Name,
+				network.NatGateway{
+					Location: to.StringPtr(s.Scope.Location()),
+					Sku:      &network.NatGatewaySku{Name: network.Standard},
+					NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
+						PublicIPAddresses: &[]network.SubResource{
+							{
+								ID: to.StringPtr(azure.PublicIPID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.NatGatewayIP.Name)),
+							},
+						},
+					},
+				},
+			)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			}
+			s.Scope.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
+		}
+	}
+	return nil
+}
+
+func (s *Service) getExisting(ctx context.Context, spec azure.NatGatewaySpec) (*infrav1.NatGateway, error) {
+	existingNatGateway, err := s.Get(ctx, s.Scope.ResourceGroup(), spec.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &infrav1.NatGateway{
+		ID:   to.String(existingNatGateway.ID),
+		Name: to.String(existingNatGateway.Name),
+		NatGatewayIP: infrav1.PublicIPSpec{
+			Name: to.String((*existingNatGateway.PublicIPAddresses)[0].ID),
```

Comment:
Is there any useful information that we're setting here that's not already in the spec? is it just the ID of the NAT Gateway?

Reply:
Yeah, only the ID of the NAT and ID of the Public IP.